crates/virtio-queue/src/mock.rs

```diff
@@ -390,30 +393,37 @@ impl<'a, M: GuestMemory> MockSplitQueue<'a, M> {
             .ok_or(Error::InvalidNextAvail)
     }
 
-    /// Adds a descriptor chain to the memory object of the queue.
-    // `descs` represents a slice of `Descriptor` objects which are used to populate the chain, and
-    // `offset` is the index in the descriptor table where the chain should be added.
+    /// Writes a single descriptor chain to the memory object of the queue, at the beginning of the
+    /// descriptor table, and returns the associated `DescriptorChain` object.
     // This method ensures the next flags and values are set properly for the desired chain, but
     // keeps the other characteristics of the input descriptors (`addr`, `len`, other flags).
-    // The descriptor chain related information is written in memory starting with address 0.
-    // The `addr` fields of the input descriptors should start at a sufficiently
-    // greater location (i.e. 1MiB, or `0x10_0000`).
-    pub fn add_desc_chain(&self, descs: &[Descriptor], offset: u16) -> Result<(), Error> {
+    pub fn build_desc_chain(&self, descs: &[Descriptor]) -> Result<DescriptorChain<&M>, Error> {
+        let mut modified_descs: Vec<Descriptor> = Vec::with_capacity(descs.len());
         for (idx, desc) in descs.iter().enumerate() {
-            let i = idx as u16 + offset;
-            let addr = desc.addr().0;
-            let len = desc.len();
             let (flags, next) = if idx == descs.len() - 1 {
                 // Clear the NEXT flag if it was set. The value of the next field of the
                 // Descriptor doesn't matter at this point.
                 (desc.flags() & !VIRTQ_DESC_F_NEXT, 0)
             } else {
                 // Ensure that the next flag is set and that we are referring the following
                 // descriptor. This ignores any value is actually present in `desc.next`.
-                (desc.flags() | VIRTQ_DESC_F_NEXT, i + 1)
+                (desc.flags() | VIRTQ_DESC_F_NEXT, idx as u16 + 1)
             };
+            modified_descs.push(Descriptor::new(desc.addr().0, desc.len(), flags, next));
+        }
+        self.build_chain_of_descriptors(&modified_descs[..])
+    }
 
-            let desc = Descriptor::new(addr, len, flags, next);
+    /// Adds descriptor chains to the memory object of the queue.
+    // `descs` represents a slice of `Descriptor` objects which are used to populate the chains, and
+    // `offset` is the index in the descriptor table where the chains should be added.
+    // The descriptor chain related information is written in memory starting with address 0.
+    // The `addr` fields of the input descriptors should start at a sufficiently
+    // greater location (i.e. 1MiB, or `0x10_0000`).
+    pub fn add_desc_chain(&self, descs: &[Descriptor], offset: u16) -> Result<(), Error> {
```

Comment:
The logic here is a bit broken. So from what I understand, this can be used to add more descriptor chains, and when the next flag is not set, I assume that will be the head of a new descriptor chain. In that case, we should add every head index into the available ring, not only the first index, and increment the avail idx with the number of new entries from the ring. Also, the function should be renamed to `add_desc_chains`.

Reply:
I changed the logic following your instructions, if I omitted something or if I misunderstood please tell me